src/benchq/resource_estimation/graph/extrapolation_estimator.py

```diff
@@ -0,0 +1,147 @@
+            "n_nodes_r_squared",
+        ]
+        inherited_necessary_info = super().__repr__() + "\n"
+
+        return inherited_necessary_info + "\n".join(
+            f"{info}: {getattr(self, info)}" for info in new_necessary_info
+        )
+
+
+class ExtrapolationResourceEstimator(GraphResourceEstimator):
+    def __init__(
+        self,
+        hw_model: BasicArchitectureModel,
+        steps_to_extrapolate_from: List[int],
+        decoder_model: Optional[DecoderModel] = None,
+        n_measurement_steps_fit_type: str = "logarithmic",
+    ):
+        self.hw_model = hw_model
+        self.steps_to_extrapolate_from = steps_to_extrapolate_from
+        self.decoder_model = decoder_model
+        self.n_measurement_steps_fit_type = n_measurement_steps_fit_type
+
+    def _get_extrapolated_graph_data(
+        self, data: List[ResourceInfo], steps_to_extrapolate_to: int
+    ) -> ExtrapolatedGraphData:
+
+        max_node_degree, max_node_degree_r_squared = _get_linear_extrapolation(
+            self.steps_to_extrapolate_from,
+            np.array([d.n_logical_qubits for d in data]),
+            steps_to_extrapolate_to,
+        )
+        # sometimes the n_measurement_steps is logarithmic, sometimes it's linear.
+        # we need to check which one is better by inspecting the fit
+        if self.n_measurement_steps_fit_type == "logarithmic":
+            (
+                n_measurement_steps,
+                n_measurement_steps_r_squared,
+            ) = _get_logarithmic_extrapolation(
+                self.steps_to_extrapolate_from,
+                np.array([d.n_measurement_steps for d in data]),
+                steps_to_extrapolate_to,
+            )
+        elif self.n_measurement_steps_fit_type == "linear":
+            (
+                n_measurement_steps,
+                n_measurement_steps_r_squared,
+            ) = _get_linear_extrapolation(
+                self.steps_to_extrapolate_from,
+                np.array([d.n_measurement_steps for d in data]),
+                steps_to_extrapolate_to,
+            )
+        else:
+            raise ValueError(
+                "n_measurement_steps_fit_type must be either 'logarithmic' or 'linear'"
+                f", not {self.n_measurement_steps_fit_type}"
+            )
+
+        n_nodes, n_nodes_r_squared = _get_linear_extrapolation(
+            self.steps_to_extrapolate_from,
+            np.array([d.n_nodes for d in data]),
+            steps_to_extrapolate_to,
+        )
+
+        return ExtrapolatedGraphData(
+            max_node_degree=max_node_degree,
+            n_measurement_steps=n_measurement_steps,
+            n_nodes=n_nodes,
+            max_node_degree_r_squared=max_node_degree_r_squared,
+            n_measurement_steps_r_squared=n_measurement_steps_r_squared,
+            n_nodes_r_squared=n_nodes_r_squared,
+        )
+
+    def estimate_via_extrapolation(
+        self,
+        data: List[ResourceInfo],
```

Comment:
I feel like `data` could be property of the class itself, so that then we could use simply `estimate` as an interface.
Though it has not been really implemented as an interface, so might be something for future.

Reply:
I don't think that would be possible. As you need to define the estimator when you call the pipeline right? And you don't have `data` until you actually run the pipeline right? Maybe the naming is a little confusing. Perhaps I should rename the variable to `data_to_extrapoate_from`

Reply:
Ok, whatever, I don't think it's worth worrying about it now, with the next refactor.